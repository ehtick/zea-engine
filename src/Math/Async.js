import { Signal } from '../Math/Signal';

class Async {
    constructor() {
        this.__asyncCount = 0;
        this.ready = new Signal(true);

        this.incAsyncCount = function(count=1) {
            this.__asyncCount+=count;
        }.bind(this);

        this.decAsyncCount = function() {
            if (this.__asyncCount > 0){
                this.__asyncCount--;
                if (this.__asyncCount == 0) {
                    this.__asyncsCompleted();
                }
            }
        }.bind(this);

        this.__asyncsCompleted = function(){
            this.ready.emit();
        }.bind(this)
    }

    get count(){
        return this.__asyncCount;
    }

};

export {
    Async
};
// export default Async;

