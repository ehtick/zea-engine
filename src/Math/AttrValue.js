// eslint-disable-next-line camelcase
import { JSON_stringify_fixedPrecision } from './Common.js'

/**
 * Base class for values that can be stored in vertex attributes.
 * <br>
 * **Note:** These values use {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Float32Array|Float32Array} values to store their data.
 */
class AttrValue {
  /**
   * Verifies that all the numeric values inside the `this.__data` variable are a valid number.
   * Returns `false` If at least one of the values is either {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Referencia/Objetos_globales/Infinity|Infinity} or
   * {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Referencia/Objetos_globales/NaN|NaN}.
   *
   * @return {boolean} - Returns the result as a boolean.
   */
  isValid() {
    for (const v of this.__data) {
      if (v == Infinity || isNaN(v)) return false
    }

    return true
  }

  /**
   * Creates a new value to wrap memory in an existing buffer.
   *
   * @param {ArrayBuffer} buffer - the buffer value.
   * @param {number} offset - the offset value.
   *
   * @todo Implement this function
   */
  static createFromFloat32Buffer(buffer, offset) {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }

  /**
   * Returns the number of Float32 elements used by this type. Used to calculate storage requirements for large arrays of this type.
   *
   * @todo Implement this function
   */
  static numElements() {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }

  /**
   * Returns current Vec2 data as array. Often used to pass types to the GPU.
   *
   * @return {array} - Returns the result as an array.
   */
  asArray() {
    return this.__data
  }

  /**
   * Calls `toJSON` method and stringifies it.
   *
   * @return {string} - The return value.
   */
  toString() {
    // eslint-disable-next-line new-cap
    return JSON_stringify_fixedPrecision(this.toJSON())
  }

  /**
   * Represents the state of your class as a JSON Object.
   *
   * @todo Implement this function in the derived class.
   */
  toJSON() {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }
}

export { AttrValue }
export default AttrValue
