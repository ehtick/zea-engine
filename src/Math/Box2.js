import {
    JSON_stringify_fixedPrecision
} from './Common.js';
import {
    Vec2
} from './Vec2';
import {
    typeRegistry
} from './TypeRegistry.js';

class Box2 {
    constructor(p0 = undefined, p1 = undefined) {
        if (p0 instanceof Vec2) {
            this.p0 = p0;
        } else {
            this.p0 = new Vec2(Number.POSITIVE_INFINITY, Number.POSITIVE_INFINITY);
        }
        if (p1 instanceof Vec2) {
            this.p1 = p1;
        } else {
            this.p1 = new Vec2(Number.NEGATIVE_INFINITY, Number.NEGATIVE_INFINITY);
        }
    }


    set(p0, p1) {
        this.p0 = p0;
        this.p1 = p1;
    }

    reset() {
        this.p0.x = Number.POSITIVE_INFINITY;
        this.p1.x = Number.NEGATIVE_INFINITY;
        this.p0.y = Number.POSITIVE_INFINITY;
        this.p1.y = Number.NEGATIVE_INFINITY;
    }

    isValid() {
        return this.p0.x != Number.POSITIVE_INFINITY && this.p1.x != Number.NEGATIVE_INFINITY &&
            this.p0.y != Number.POSITIVE_INFINITY && this.p1.y != Number.NEGATIVE_INFINITY;
    }

    addPoint(point) {
        if (this.p0.x == Number.POSITIVE_INFINITY || point.x < this.p0.x)
            this.p0.x = point.x;
        if (this.p0.y == Number.POSITIVE_INFINITY || point.y < this.p0.y)
            this.p0.y = point.y;

        if (this.p1.y == Number.NEGATIVE_INFINITY || point.x > this.p1.x)
            this.p1.x = point.x;
        if (this.p1.y == Number.NEGATIVE_INFINITY || point.y > this.p1.y)
            this.p1.y = point.y;
    }

    size() {
        return this.p1.subtract(this.p0);
    }
    diagonal() {
        return this.p1.subtract(this.p0);
    }

    center() {
        let result = this.p1.subtract(this.p0);
        result.scaleInPlace(0.5);
        result.addInPlace(this.p0);
        return result;
    }


    //////////////////////////////////////////
    // Static Methods

    static create(...args) {
        return new Box2(...args);
    }

    //////////////////////////////////////////
    // Persistence

    toJSON() {
        return {
            "p0": this.p0.toJSON(),
            "p1": this.p1.toJSON()
        }
    }

    toString() {
        return JSON_stringify_fixedPrecision(this.toJSON())
    }
};

typeRegistry.registerType('Box2', Box2);

export {
    Box2
};
// export default Box2;