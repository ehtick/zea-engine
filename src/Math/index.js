
export {
    SInt32,
    UInt32,
    Float32,
    hashStr,
    JSON_stringify_fixedPrecision
} from './Common.js';
export * from './AttrValue.js';
export * from './Vec2.js';
export * from './Vec3.js';
export * from './Vec4.js';
export * from './Color.js';
export * from './Quat.js';
export * from './Ray.js';
export * from './Mat3.js';
export * from './Mat4.js';
export * from './Xfo.js';
export * from './Box2.js';
export * from './Box3.js';
export * from './Rect2.js';
export * from './Hammersley.js';

export * from './TypeRegistry.js';
