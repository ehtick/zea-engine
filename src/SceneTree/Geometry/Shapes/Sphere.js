import { Vec2 } from '../../../Math/Vec2';
import { Vec3 } from '../../../Math/Vec3';
import { Mesh } from '../Mesh.js';


class Sphere extends Mesh {
    constructor(radius = 1.0, sides = 12, loops = 12) {
        super();

        if(isNaN(radius) || isNaN(sides) || isNaN(loops))
            throw("Invalid geom args");

        this.__radius = radius;
        this.__sides = (sides >= 3) ? sides : 3;
        this.__loops = (loops >= 3) ? loops : 3;

        this.addVertexAttribute('texCoords', Vec2);
        this.addVertexAttribute('normals', Vec3);
        this.__rebuild();
    }

    get radius() {
        return this.__radius
    }

    set radius(val) {
        this.__radius = val;
        this.__resize();
    }

    get sides() {
        return this.__sides
    }

    set sides(val) {
        this.__sides = (val >= 3) ? val : 3;
        this.__rebuild();
    }

    get loops() {
        return this.__loops
    }

    set loops(val) {
        this.__loops = (val >= 3) ? val : 3;
        this.__rebuild();
    }

    __rebuild() {
        let numVertices = 2 + this.__sides * this.__loops;
        let numTris = this.__sides * 2;
        let numQuads = this.__sides * this.__loops;
        this.setNumVertices(numVertices);
        this.setFaceCounts([numTris, numQuads]);

        //////////////////////////////
        // Set Vertex Positions

        let normals = this.getVertexAttribute('normals');
        let vertex = 0;
        let normal = new Vec3(0.0, 0.0, 1.0);
        this.getVertex(vertex).set(0.0, 0.0, this.__radius);
        normals.getValueRef(vertex).set(0.0, 0.0, 1.0);
        vertex++;

        for (let i = 0; i < this.__loops; i++) {
            let theta = ((i + 1) / (this.__loops + 1)) * Math.PI;
            for (let j = 0; j < this.__sides; j++) {
                let phi = (j / this.__sides) * 2.0 * Math.PI;
                normal.set(Math.sin(theta) * Math.cos(phi), Math.sin(theta) * Math.sin(phi), Math.cos(theta));

                // Set positions and normals at the same time.
                this.getVertex(vertex).setFromOther(normal.scale(this.__radius));
                normals.getValueRef(vertex).setFromOther(normal);
                vertex++;
            }
        }
        this.getVertex(vertex).set(0.0, 0.0, -this.__radius);
        normals.getValueRef(vertex).set(0.0, 0.0, -1.0);
        vertex++;

        //////////////////////////////
        // build the topology
        let texCoords = this.getVertexAttribute('texCoords');

        // build the fan at the first pole.
        let faceIndex = 0;
        for (let j = 0; j < this.__sides; j++) {
            let v0 = 0;
            let v1 = ((j + 1) % this.__sides) + 1;
            let v2 = j + 1;
            this.setFaceVertexIndices(faceIndex, v0, v1, v2);

            let uv0 = new Vec2(0.5, 0.0);
            let uv1 = new Vec2(1.0 - ((j + 1) / this.__sides), 0.0);
            let uv2 = new Vec2(1.0 - (j / this.__sides), 1.0 / (this.__loops + 1));
            texCoords.setFaceVertexValue(faceIndex, 0, uv0);
            texCoords.setFaceVertexValue(faceIndex, 1, uv1);
            texCoords.setFaceVertexValue(faceIndex, 2, uv2);

            faceIndex++;
        }
        // build the fan at the second pole.
        for (let j = 0; j < this.__sides; j++) {
            let v0 = numVertices - 1;
            let v1 = (this.__sides * (this.__loops - 1)) + j + 1;
            let v2 = (this.__sides * (this.__loops - 1)) + ((j + 1) % this.__sides) + 1;
            this.setFaceVertexIndices(faceIndex, v0, v1, v2);

            let uv0 = new Vec2(1.0 - (j / this.__sides), this.__loops / (this.__loops + 1));
            let uv1 = new Vec2(1.0 - ((j + 1) / this.__sides), this.__loops / (this.__loops + 1));
            let uv2 = new Vec2(0.5, 1.0);
            texCoords.setFaceVertexValue(faceIndex, 0, uv0);
            texCoords.setFaceVertexValue(faceIndex, 1, uv1);
            texCoords.setFaceVertexValue(faceIndex, 2, uv2);

            faceIndex++;
        }

        for (let i = 0; i < this.__loops - 1; i++) {
            for (let j = 0; j < this.__sides; j++) {
                let v0 = (this.__sides * i) + j + 1;
                let v1 = (this.__sides * i) + ((j + 1) % this.__sides) + 1;
                let v2 = (this.__sides * (i + 1)) + ((j + 1) % this.__sides) + 1;
                let v3 = (this.__sides * (i + 1)) + j + 1;
                this.setFaceVertexIndices(faceIndex, v0, v1, v2, v3);

                texCoords.setFaceVertexValue(faceIndex, 0, new Vec2(i / this.__loops, j / this.__loops));
                texCoords.setFaceVertexValue(faceIndex, 1, new Vec2(i / this.__loops, (j + 1) / this.__loops));
                texCoords.setFaceVertexValue(faceIndex, 2, new Vec2((i + 1) / this.__loops, (j + 1) / this.__loops));
                texCoords.setFaceVertexValue(faceIndex, 3, new Vec2((i + 1) / this.__loops, j / this.__loops));
                faceIndex++;
            }
        }

        this.setBoundingBoxDirty();
    }

    __resize() {
        //////////////////////////////
        // Set Vertex Positions
        let vertex = 0;
        let normal = new Vec3(0.0, 0.0, 1.0);
        this.getVertex(vertex).set(0.0, 0.0, this.__radius);
        vertex++;

        for (let i = 0; i < this.__loops; i++) {
            let theta = ((i + 1) / (this.__loops + 1)) * Math.PI;
            for (let j = 0; j < this.__sides; j++) {
                let phi = (j / this.__sides) * 2.0 * Math.PI;
                normal.set(Math.sin(theta) * Math.cos(phi), Math.sin(theta) * Math.sin(phi), Math.cos(theta));

                // Set positions and normals at the same time.
                this.getVertex(vertex).setFromOther(normal.scale(this.__radius));
                vertex++;
            }
        }
        this.getVertex(vertex).set(0.0, 0.0, -this.__radius);
        vertex++;

        this.setBoundingBoxDirty();
    }

    toJSON() {
        let json = super.toJSON();
        json['x'] = this.__x;
        json['y'] = this.__y;
        json['z'] = this.__z;
        return json
    }
};

export {
    Sphere
};
//export default Sphere;