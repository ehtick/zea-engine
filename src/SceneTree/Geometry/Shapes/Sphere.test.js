import { Sphere } from './Sphere'
import { Vec3 } from '../../../Math'

describe('Sphere', () => {
  test('test default parameters.', () => {
    const sphere = new Sphere()

    expect(sphere.getParameter('Radius').getValue()).toBe(1.0)
    expect(sphere.getParameter('Sides').getValue()).toBe(12)
    expect(sphere.getParameter('Loops').getValue()).toBe(12)
  })

  test('test update parameters.', () => {
    const sphere = new Sphere()
    sphere.getParameter('Radius').setValue(2.6)
    sphere.getParameter('Sides').setValue(24)
    sphere.getParameter('Loops').setValue(30)

    expect(sphere.getParameter('Radius').getValue()).toBe(2.6)
    expect(sphere.getParameter('Sides').getValue()).toBe(24)
    expect(sphere.getParameter('Loops').getValue()).toBe(30)
  })

  it('saves to JSON (serialization).', () => {
    const sphere = new Sphere(2.6, 5, 5)
    const outputJSON = sphere.toJSON()

    expect(outputJSON).toMatchSnapshot()
  })

  it('restores from JSON (serialization).', () => {
    const sphere = new Sphere()
    const inputStr =
      '{"params":{"Radius":{"value":2.6},"Sides":{"value":5,"range":[3,200],"step":1},"Loops":{"value":5,"range":[3,200],"step":1}},"type":"Sphere","numVertices":27,"vertexAttributes":{"positions":{"data":[0,0,2.5999999046325684,1.2999999523162842,0,2.2516660690307617,0.40172210335731506,1.2363735437393188,2.2516660690307617,-1.0517220497131348,0.7641208171844482,2.2516660690307617,-1.0517220497131348,-0.7641208171844482,2.2516660690307617,0.40172210335731506,-1.2363735437393188,2.2516660690307617,2.2516660690307617,0,1.2999999523162842,0.6958031058311462,2.1414616107940674,1.2999999523162842,-1.8216361999511719,1.3234961032867432,1.2999999523162842,-1.8216361999511719,-1.3234961032867432,1.2999999523162842,0.6958031058311462,-2.1414616107940674,1.2999999523162842,2.5999999046325684,0,1.5920409083607182e-16,0.8034442067146301,2.4727470874786377,1.5920409083607182e-16,-2.1034440994262695,1.5282416343688965,1.5920409083607182e-16,-2.1034440994262695,-1.5282416343688965,1.5920409083607182e-16,0.8034442067146301,-2.4727470874786377,1.5920409083607182e-16,2.2516660690307617,0,-1.2999999523162842,0.6958031058311462,2.1414616107940674,-1.2999999523162842,-1.8216361999511719,1.3234961032867432,-1.2999999523162842,-1.8216361999511719,-1.3234961032867432,-1.2999999523162842,0.6958031058311462,-2.1414616107940674,-1.2999999523162842,1.2999999523162842,0,-2.2516660690307617,0.40172210335731506,1.2363735437393188,-2.2516660690307617,-1.0517220497131348,0.7641208171844482,-2.2516660690307617,-1.0517220497131348,-0.7641208171844482,-2.2516660690307617,0.40172210335731506,-1.2363735437393188,-2.2516660690307617,0,0,-2.5999999046325684],"dataType":"Vec3","defaultValue":0,"length":27,"splits":{},"splitValues":[]},"texCoords":{"data":[0.5,0,1,0.1666666716337204,0.800000011920929,0,0.6000000238418579,0,0.4000000059604645,0,0.20000000298023224,0,0.20000000298023224,0,0.20000000298023224,0.20000000298023224,0.20000000298023224,0.4000000059604645,0.20000000298023224,0.6000000238418579,0.20000000298023224,0.800000011920929,0.4000000059604645,0,0.4000000059604645,0.20000000298023224,0.4000000059604645,0.4000000059604645,0.4000000059604645,0.6000000238418579,0.4000000059604645,0.800000011920929,0.6000000238418579,0,0.6000000238418579,0.20000000298023224,0.6000000238418579,0.4000000059604645,0.6000000238418579,0.6000000238418579,0.6000000238418579,0.800000011920929,0.800000011920929,0.8333333134651184,0.5,1,0.5,1,0.5,1,0.5,1,1,0.8333333134651184],"dataType":"Vec2","defaultValue":1.7976931348623157e+308,"length":27,"splits":{"1":{"4":3,"10":3,"14":18},"2":{"1":0,"10":14,"11":14},"3":{"2":1,"11":15,"12":15},"4":{"3":2,"12":16,"13":16},"5":{"4":4,"13":17,"14":17},"6":{"14":19,"19":19},"11":{"19":20,"24":20},"16":{"24":21,"29":21},"21":{"9":13,"25":23,"29":27},"22":{"6":6,"25":22,"26":22},"23":{"7":8,"26":24,"27":24},"24":{"8":10,"27":25,"28":25},"25":{"9":12,"28":26,"29":26},"26":{"6":5,"7":7,"8":9,"9":11}},"splitValues":[{"x":0.800000011920929,"y":0.1666666716337204},{"x":0.6000000238418579,"y":0.1666666716337204},{"x":0.4000000059604645,"y":0.1666666716337204},{"x":0,"y":0},{"x":0.20000000298023224,"y":0.1666666716337204},{"x":0.800000011920929,"y":0.8333333134651184},{"x":0.6000000238418579,"y":0.8333333134651184},{"x":0.6000000238418579,"y":0.8333333134651184},{"x":0.4000000059604645,"y":0.8333333134651184},{"x":0.4000000059604645,"y":0.8333333134651184},{"x":0.20000000298023224,"y":0.8333333134651184},{"x":0.20000000298023224,"y":0.8333333134651184},{"x":0,"y":0.8333333134651184},{"x":0.5,"y":1},{"x":0,"y":0.20000000298023224},{"x":0,"y":0.4000000059604645},{"x":0,"y":0.6000000238418579},{"x":0,"y":0.800000011920929},{"x":0,"y":1},{"x":0.20000000298023224,"y":1},{"x":0.4000000059604645,"y":1},{"x":0.6000000238418579,"y":1},{"x":0.800000011920929,"y":0.20000000298023224},{"x":0.800000011920929,"y":0},{"x":0.800000011920929,"y":0.4000000059604645},{"x":0.800000011920929,"y":0.6000000238418579},{"x":0.800000011920929,"y":0.800000011920929},{"x":0.800000011920929,"y":1}]},"normals":{"data":[0,0,1,0.5,0,0.8660253882408142,0.15450850129127502,0.4755282700061798,0.8660253882408142,-0.404508501291275,0.29389262199401855,0.8660253882408142,-0.404508501291275,-0.29389262199401855,0.8660253882408142,0.15450850129127502,-0.4755282700061798,0.8660253882408142,0.8660253882408142,0,0.5,0.2676165699958801,0.8236390948295593,0.5,-0.7006292939186096,0.5090369582176208,0.5,-0.7006292939186096,-0.5090369582176208,0.5,0.2676165699958801,-0.8236390948295593,0.5,1,0,6.123234262925839e-17,0.30901700258255005,0.9510565400123596,6.123234262925839e-17,-0.80901700258255,0.5877852439880371,6.123234262925839e-17,-0.80901700258255,-0.5877852439880371,6.123234262925839e-17,0.30901700258255005,-0.9510565400123596,6.123234262925839e-17,0.8660253882408142,0,-0.5,0.2676165699958801,0.8236390948295593,-0.5,-0.7006292939186096,0.5090369582176208,-0.5,-0.7006292939186096,-0.5090369582176208,-0.5,0.2676165699958801,-0.8236390948295593,-0.5,0.5,0,-0.8660253882408142,0.15450850129127502,0.4755282700061798,-0.8660253882408142,-0.404508501291275,0.29389262199401855,-0.8660253882408142,-0.404508501291275,-0.29389262199401855,-0.8660253882408142,0.15450850129127502,-0.4755282700061798,-0.8660253882408142,0,0,-1],"dataType":"Vec3","defaultValue":1.7976931348623157e+308,"length":27,"splits":{},"splitValues":[]}},"faceCounts":[10,25],"faceVertexIndices":[0,2,1,0,3,2,0,4,3,0,5,4,0,1,5,26,21,22,26,22,23,26,23,24,26,24,25,26,25,21,1,2,7,6,2,3,8,7,3,4,9,8,4,5,10,9,5,1,6,10,6,7,12,11,7,8,13,12,8,9,14,13,9,10,15,14,10,6,11,15,11,12,17,16,12,13,18,17,13,14,19,18,14,15,20,19,15,11,16,20,16,17,22,21,17,18,23,22,18,19,24,23,19,20,25,24,20,16,21,25,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}'
    sphere.fromJSON(JSON.parse(inputStr))

    expect(sphere.toJSON()).toEqual(new Sphere(2.6, 5, 5).toJSON())
  })
})
