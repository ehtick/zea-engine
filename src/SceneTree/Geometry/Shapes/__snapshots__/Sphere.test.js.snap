// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sphere saves to JSON (serialization). 1`] = `
Object {
  "faceCounts": Array [
    10,
    25,
  ],
  "faceVertexIndices": Array [
    0,
    2,
    1,
    0,
    3,
    2,
    0,
    4,
    3,
    0,
    5,
    4,
    0,
    1,
    5,
    26,
    21,
    22,
    26,
    22,
    23,
    26,
    23,
    24,
    26,
    24,
    25,
    26,
    25,
    21,
    1,
    2,
    7,
    6,
    2,
    3,
    8,
    7,
    3,
    4,
    9,
    8,
    4,
    5,
    10,
    9,
    5,
    1,
    6,
    10,
    6,
    7,
    12,
    11,
    7,
    8,
    13,
    12,
    8,
    9,
    14,
    13,
    9,
    10,
    15,
    14,
    10,
    6,
    11,
    15,
    11,
    12,
    17,
    16,
    12,
    13,
    18,
    17,
    13,
    14,
    19,
    18,
    14,
    15,
    20,
    19,
    15,
    11,
    16,
    20,
    16,
    17,
    22,
    21,
    17,
    18,
    23,
    22,
    18,
    19,
    24,
    23,
    19,
    20,
    25,
    24,
    20,
    16,
    21,
    25,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "numVertices": 27,
  "params": Object {
    "Loops": Object {
      "range": Array [
        3,
        200,
      ],
      "step": 1,
      "value": 5,
    },
    "Radius": Object {
      "value": 2.6,
    },
    "Sides": Object {
      "range": Array [
        3,
        200,
      ],
      "step": 1,
      "value": 5,
    },
  },
  "type": "Sphere",
  "vertexAttributes": Object {
    "normals": Object {
      "data": Array [
        0,
        0,
        1,
        0.5,
        0,
        0.8660253882408142,
        0.15450850129127502,
        0.4755282700061798,
        0.8660253882408142,
        -0.404508501291275,
        0.29389262199401855,
        0.8660253882408142,
        -0.404508501291275,
        -0.29389262199401855,
        0.8660253882408142,
        0.15450850129127502,
        -0.4755282700061798,
        0.8660253882408142,
        0.8660253882408142,
        0,
        0.5,
        0.2676165699958801,
        0.8236390948295593,
        0.5,
        -0.7006292939186096,
        0.5090369582176208,
        0.5,
        -0.7006292939186096,
        -0.5090369582176208,
        0.5,
        0.2676165699958801,
        -0.8236390948295593,
        0.5,
        1,
        0,
        6.123234262925839e-17,
        0.30901700258255005,
        0.9510565400123596,
        6.123234262925839e-17,
        -0.80901700258255,
        0.5877852439880371,
        6.123234262925839e-17,
        -0.80901700258255,
        -0.5877852439880371,
        6.123234262925839e-17,
        0.30901700258255005,
        -0.9510565400123596,
        6.123234262925839e-17,
        0.8660253882408142,
        0,
        -0.5,
        0.2676165699958801,
        0.8236390948295593,
        -0.5,
        -0.7006292939186096,
        0.5090369582176208,
        -0.5,
        -0.7006292939186096,
        -0.5090369582176208,
        -0.5,
        0.2676165699958801,
        -0.8236390948295593,
        -0.5,
        0.5,
        0,
        -0.8660253882408142,
        0.15450850129127502,
        0.4755282700061798,
        -0.8660253882408142,
        -0.404508501291275,
        0.29389262199401855,
        -0.8660253882408142,
        -0.404508501291275,
        -0.29389262199401855,
        -0.8660253882408142,
        0.15450850129127502,
        -0.4755282700061798,
        -0.8660253882408142,
        0,
        0,
        -1,
      ],
      "dataType": "Vec3",
      "defaultValue": 1.7976931348623157e+308,
      "length": 27,
      "splitValues": Array [],
      "splits": Object {},
    },
    "positions": Object {
      "data": Array [
        0,
        0,
        2.5999999046325684,
        1.2999999523162842,
        0,
        2.2516660690307617,
        0.40172210335731506,
        1.2363735437393188,
        2.2516660690307617,
        -1.0517220497131348,
        0.7641208171844482,
        2.2516660690307617,
        -1.0517220497131348,
        -0.7641208171844482,
        2.2516660690307617,
        0.40172210335731506,
        -1.2363735437393188,
        2.2516660690307617,
        2.2516660690307617,
        0,
        1.2999999523162842,
        0.6958031058311462,
        2.1414616107940674,
        1.2999999523162842,
        -1.8216361999511719,
        1.3234961032867432,
        1.2999999523162842,
        -1.8216361999511719,
        -1.3234961032867432,
        1.2999999523162842,
        0.6958031058311462,
        -2.1414616107940674,
        1.2999999523162842,
        2.5999999046325684,
        0,
        1.5920409083607182e-16,
        0.8034442067146301,
        2.4727470874786377,
        1.5920409083607182e-16,
        -2.1034440994262695,
        1.5282416343688965,
        1.5920409083607182e-16,
        -2.1034440994262695,
        -1.5282416343688965,
        1.5920409083607182e-16,
        0.8034442067146301,
        -2.4727470874786377,
        1.5920409083607182e-16,
        2.2516660690307617,
        0,
        -1.2999999523162842,
        0.6958031058311462,
        2.1414616107940674,
        -1.2999999523162842,
        -1.8216361999511719,
        1.3234961032867432,
        -1.2999999523162842,
        -1.8216361999511719,
        -1.3234961032867432,
        -1.2999999523162842,
        0.6958031058311462,
        -2.1414616107940674,
        -1.2999999523162842,
        1.2999999523162842,
        0,
        -2.2516660690307617,
        0.40172210335731506,
        1.2363735437393188,
        -2.2516660690307617,
        -1.0517220497131348,
        0.7641208171844482,
        -2.2516660690307617,
        -1.0517220497131348,
        -0.7641208171844482,
        -2.2516660690307617,
        0.40172210335731506,
        -1.2363735437393188,
        -2.2516660690307617,
        0,
        0,
        -2.5999999046325684,
      ],
      "dataType": "Vec3",
      "defaultValue": 0,
      "length": 27,
      "splitValues": Array [],
      "splits": Object {},
    },
    "texCoords": Object {
      "data": Array [
        0.5,
        0,
        1,
        0.1666666716337204,
        0.800000011920929,
        0,
        0.6000000238418579,
        0,
        0.4000000059604645,
        0,
        0.20000000298023224,
        0,
        0.20000000298023224,
        0,
        0.20000000298023224,
        0.20000000298023224,
        0.20000000298023224,
        0.4000000059604645,
        0.20000000298023224,
        0.6000000238418579,
        0.20000000298023224,
        0.800000011920929,
        0.4000000059604645,
        0,
        0.4000000059604645,
        0.20000000298023224,
        0.4000000059604645,
        0.4000000059604645,
        0.4000000059604645,
        0.6000000238418579,
        0.4000000059604645,
        0.800000011920929,
        0.6000000238418579,
        0,
        0.6000000238418579,
        0.20000000298023224,
        0.6000000238418579,
        0.4000000059604645,
        0.6000000238418579,
        0.6000000238418579,
        0.6000000238418579,
        0.800000011920929,
        0.800000011920929,
        0.8333333134651184,
        0.5,
        1,
        0.5,
        1,
        0.5,
        1,
        0.5,
        1,
        1,
        0.8333333134651184,
      ],
      "dataType": "Vec2",
      "defaultValue": 1.7976931348623157e+308,
      "length": 27,
      "splitValues": Array [
        Object {
          "x": 0.800000011920929,
          "y": 0.1666666716337204,
        },
        Object {
          "x": 0.6000000238418579,
          "y": 0.1666666716337204,
        },
        Object {
          "x": 0.4000000059604645,
          "y": 0.1666666716337204,
        },
        Object {
          "x": 0,
          "y": 0,
        },
        Object {
          "x": 0.20000000298023224,
          "y": 0.1666666716337204,
        },
        Object {
          "x": 0.800000011920929,
          "y": 0.8333333134651184,
        },
        Object {
          "x": 0.6000000238418579,
          "y": 0.8333333134651184,
        },
        Object {
          "x": 0.6000000238418579,
          "y": 0.8333333134651184,
        },
        Object {
          "x": 0.4000000059604645,
          "y": 0.8333333134651184,
        },
        Object {
          "x": 0.4000000059604645,
          "y": 0.8333333134651184,
        },
        Object {
          "x": 0.20000000298023224,
          "y": 0.8333333134651184,
        },
        Object {
          "x": 0.20000000298023224,
          "y": 0.8333333134651184,
        },
        Object {
          "x": 0,
          "y": 0.8333333134651184,
        },
        Object {
          "x": 0.5,
          "y": 1,
        },
        Object {
          "x": 0,
          "y": 0.20000000298023224,
        },
        Object {
          "x": 0,
          "y": 0.4000000059604645,
        },
        Object {
          "x": 0,
          "y": 0.6000000238418579,
        },
        Object {
          "x": 0,
          "y": 0.800000011920929,
        },
        Object {
          "x": 0,
          "y": 1,
        },
        Object {
          "x": 0.20000000298023224,
          "y": 1,
        },
        Object {
          "x": 0.4000000059604645,
          "y": 1,
        },
        Object {
          "x": 0.6000000238418579,
          "y": 1,
        },
        Object {
          "x": 0.800000011920929,
          "y": 0.20000000298023224,
        },
        Object {
          "x": 0.800000011920929,
          "y": 0,
        },
        Object {
          "x": 0.800000011920929,
          "y": 0.4000000059604645,
        },
        Object {
          "x": 0.800000011920929,
          "y": 0.6000000238418579,
        },
        Object {
          "x": 0.800000011920929,
          "y": 0.800000011920929,
        },
        Object {
          "x": 0.800000011920929,
          "y": 1,
        },
      ],
      "splits": Object {
        "1": Object {
          "10": 3,
          "14": 18,
          "4": 3,
        },
        "11": Object {
          "19": 20,
          "24": 20,
        },
        "16": Object {
          "24": 21,
          "29": 21,
        },
        "2": Object {
          "1": 0,
          "10": 14,
          "11": 14,
        },
        "21": Object {
          "25": 23,
          "29": 27,
          "9": 13,
        },
        "22": Object {
          "25": 22,
          "26": 22,
          "6": 6,
        },
        "23": Object {
          "26": 24,
          "27": 24,
          "7": 8,
        },
        "24": Object {
          "27": 25,
          "28": 25,
          "8": 10,
        },
        "25": Object {
          "28": 26,
          "29": 26,
          "9": 12,
        },
        "26": Object {
          "6": 5,
          "7": 7,
          "8": 9,
          "9": 11,
        },
        "3": Object {
          "11": 15,
          "12": 15,
          "2": 1,
        },
        "4": Object {
          "12": 16,
          "13": 16,
          "3": 2,
        },
        "5": Object {
          "13": 17,
          "14": 17,
          "4": 4,
        },
        "6": Object {
          "14": 19,
          "19": 19,
        },
      },
    },
  },
}
`;
